jobqueue/mysql/inspector.go

```diff
@@ -36,37 +36,142 @@ func (i *inspector) Find(jobID uint64) (*jobqueue.InspectedJob, error) {
 	return j, nil
 }
 
-func (i *inspector) FindAllGrabbed(limit uint, cursor string) (*jobqueue.InspectedJobs, error) {
+func (i *inspector) FindAllGrabbed(limit uint, cursor string, order jobqueue.SortOrder) (*jobqueue.InspectedJobs, error) {
 	var maxTime = time.Now().UnixNano() / int64(time.Millisecond)
-	return i.findAll("grabbed", 0, maxTime, limit, cursor)
+	if order == jobqueue.Asc {
+		return i.findAllAsc("grabbed", 0, maxTime, limit, cursor)
+	}
+	return i.findAllDesc("grabbed", 0, maxTime, limit, cursor)
 }
 
-func (i *inspector) FindAllWaiting(limit uint, cursor string) (*jobqueue.InspectedJobs, error) {
+func (i *inspector) FindAllWaiting(limit uint, cursor string, order jobqueue.SortOrder) (*jobqueue.InspectedJobs, error) {
 	var maxTime = time.Now().UnixNano() / int64(time.Millisecond)
-	return i.findAll("claimed", 0, maxTime, limit, cursor)
+	if order == jobqueue.Asc {
+		return i.findAllAsc("claimed", 0, maxTime, limit, cursor)
+	}
+	return i.findAllDesc("claimed", 0, maxTime, limit, cursor)
 }
 
-func (i *inspector) FindAllDeferred(limit uint, cursor string) (*jobqueue.InspectedJobs, error) {
+func (i *inspector) FindAllDeferred(limit uint, cursor string, order jobqueue.SortOrder) (*jobqueue.InspectedJobs, error) {
 	var minTime = time.Now().UnixNano() / int64(time.Millisecond)
-	return i.findAll("claimed", minTime, 0, limit, cursor)
+	if order == jobqueue.Asc {
+		return i.findAllAsc("claimed", minTime, math.MaxInt64, limit, cursor)
+	}
+	return i.findAllDesc("claimed", minTime, 0, limit, cursor)
 }
 
-func (i *inspector) findAll(status string, minTime int64, maxTime int64, limit uint, cursor string) (*jobqueue.InspectedJobs, error) {
-	if maxTime <= 0 {
-		maxTime = math.MaxInt64
-	}
-	var maxJobID uint64 = math.MaxUint64
+func decodeCursor(cursor string, time *int64, jobID *uint64) {
 	if decoded, err := base64.StdEncoding.DecodeString(cursor); err == nil {
 		if pair := strings.SplitN(string(decoded), ",", 2); len(pair) == 2 {
 			t, err1 := strconv.Atoi(pair[0])
 			j, err2 := strconv.Atoi(pair[1])
 			if err1 == nil && err2 == nil {
-				maxTime = int64(t)
-				maxJobID = uint64(j)
+				*time = int64(t)
+				*jobID = uint64(j)
+			}
+		}
+	}
+}
+
+func (i *inspector) findAllAsc(status string, minTime int64, maxTime int64, limit uint, cursor string) (*jobqueue.InspectedJobs, error) {
```

Comment:
This method has many duplicated code with `findAllDesc` 🤔  but I cannot split these method as simple.

Reply:
It doesn't surprise me.  There is duplicated code already in `failureLog`.